### ISSUE TITLE
[v1.3.0 regression] Nerdctl push with Cosign not working with real registries

### ISSUE THREAD

**Issue by ningziwen** [containerd/nerdctl#2171]
### Description

Nerdctl push with Cosign seems not working with real registries (like Docker Hub or ECR) for me with the latest nerdctl, although the local registry may still work because it is part of nerdctl CI.

With the latest nerdctl:
```
sudo -E COSIGN_PASSWORD=123 ./_output/nerdctl push --sign cosign --cosign-key cosign.key charlienzw/sign:tagname --debug

...
DEBU[0000] running /home/ubuntu/go/bin/cosign [/home/ubuntu/go/bin/cosign sign --key cosign.key --yes docker.io/charlienzw/sign:tagname@sha256:4e83453afed1b4fa1a3500525091dbfca6ce1e66903fd4c01ff015dbcb1ba33e] 
INFO[0000] cosign: Error: signing [docker.io/charlienzw/sign:tagname@sha256:4e83453afed1b4fa1a3500525091dbfca6ce1e66903fd4c01ff015dbcb1ba33e]: accessing image: entity not found in registry 
INFO[0000] cosign: main.go:74: error during command execution: signing [docker.io/charlienzw/sign:tagname@sha256:4e83453afed1b4fa1a3500525091dbfca6ce1e66903fd4c01ff015dbcb1ba33e]: accessing image: entity not found in registry 
```

Reverted this [commit](https://github.com/containerd/nerdctl/pull/2136) locally and it works and the related debug log is 
```
DEBU[0000] running /home/ubuntu/go/bin/cosign [/home/ubuntu/go/bin/cosign sign --key cosign.key charlienzw/sign:tagname]
```

Both of the experiments are using Cosign 2.0.0.

Looks like the reason is that the latest nerdctl adds the digest at the end of image before passing to Cosign, and the digest added is from local, which is different with the one in Dockerhub/ECR (because container registry compresses the image which changes the digest displayed in registry, and local registry may not have this concern).

### Steps to reproduce the issue

Build nerdctl from main or use 1.3.0, then follow https://github.com/containerd/nerdctl/blob/main/docs/cosign.md to push a image with test key to Docker Hub, with Cosign 2.0 installed.

### Describe the results you received and expected

I expect the latest version of nerdctl also succeeds for pushing with Cosign.

### What version of nerdctl are you using?

1.3.0 or main branch can reproduce.

Reverting the [commit](https://github.com/containerd/nerdctl/pull/2136) can work.

### Are you using a variant of nerdctl? (e.g., Rancher Desktop)

None

### Host information

```
Admin:~/nerdctl (main) $ ./_output/nerdctl info
Client:
 Namespace:     default
 Debug Mode:    false

Server:
 Server Version: 1.6.16
 Storage Driver: overlayfs
 Logging Driver: json-file
 Cgroup Driver: none
 Cgroup Version: 1
 Plugins:
  Log: fluentd journald json-file syslog
  Storage: aufs native overlayfs
 Security Options:
  apparmor
  seccomp
   Profile: default
  rootless
 Kernel Version: 5.4.0-1099-aws
 Operating System: Ubuntu 18.04.6 LTS
 OSType: linux
 Architecture: x86_64
 CPUs: 2
 Total Memory: 7.483GiB
 Name: ip-172-31-84-158
 ID: 9c5e3b7f-301a-48cf-82f3-133f608b8185

WARNING: AppArmor profile "nerdctl-default" is not loaded.
         Use 'sudo nerdctl apparmor load' if you prefer to use AppArmor with rootless mode.
         This warning is negligible if you do not intend to use AppArmor.
WARNING: Running in rootless-mode without cgroups. To enable cgroups in rootless-mode, you need to boot the system in cgroup v2 mode.
```

**Comment by ningziwen** [comment]
cc @developer-guy @AkihiroSuda @djdongjin to discuss the change. https://github.com/containerd/nerdctl/pull/2136

**Comment by ningziwen** [comment]
Reading through the [issue](https://github.com/containerd/nerdctl/issues/2135), I understand the concern of race condition. However, as it is common that the digest doesn't match in registry and local (if this assumption is correct), nerdctl has to get the digest to be used by Cosign from the registry. However inserting a HEAD request between push and sign still has the race condition, and I can't think other ways of getting the digest in registry.

So unless anyone has better idea, we might need to acknowledge the race condition risk here and add a word in documentation to ask users to take their own risk if they want to push with sign using image tag.

**Comment by AkihiroSuda** [comment]
Thanks for reporting.

> the digest added is from local, which is different with the one in Dockerhub/ECR (because container registry compresses the image which changes the digest displayed in registry, and local registry may not have this concern).

I doubt this.
I assume that the digest is changed just due to a nerdctl issue related to multi-platform.

**Comment by ningziwen** [comment]
> Thanks for reporting.
> 
> > the digest added is from local, which is different with the one in Dockerhub/ECR (because container registry compresses the image which changes the digest displayed in registry, and local registry may not have this concern).
> 
> I doubt this. I assume that the digest is changed just due to a nerdctl issue related to multi-platform.

Did a second look and you are right :)

There may still be some unknown gaps between distribution and the real registries, which can explain why it escaped the [test](https://github.com/containerd/nerdctl/blob/11b1fc2da89727fb8fdf496543ef7a5294a9dedd/cmd/nerdctl/image_pull_linux_test.go#L87).

**Comment by AkihiroSuda** [comment]
Fixed in #2172
